aeneas/src/x86/X86MacroAssembler.v3

```diff
@@ -51,22 +73,51 @@ class X86MacroAssembler extends X86Assembler {
 		push(s);            // push value from source memory onto the stack
 		pop(X86Addr.!(d));  // pop value off stack into destination memory
 	}
+	// a macro to move between any two SSE register/memory operands
+	def movsd_sm_sm(d: SSERm, s: SSERm, scratch: SSEReg) {
+		if (s == d) return;
+		if (SSEReg.?(d)) return movsd_s_sm(SSEReg.!(d), s);
+		if (SSEReg.?(s)) return movsd_sm_s(d, SSEReg.!(s));
+		if (scratch != null) {
+			movsd_s_sm(scratch, s);
+			movsd_sm_s(d, scratch);
+		}
+		// push(s);            // push value from source memory onto the stack
+		// pop(SSEAddr.!(d));  // pop value off stack into destination memory
+	}
 	// a macro to move a value into a location
 	def movd_l_val(frame: MachFrame, loc: int, val: Val) {
-		var d = loc_rm(frame, loc);
-		if (val == null) {
-			movd_rm_i(d, 0);
-		} else if (Addr.?(val)) {
-			var pos = buffer.pos;
-			movd_rm_i(d, X86Addrs.ABS_CONST);
-			recordPatch(pos, Addr.!(val));
+		if (loc >= X86MachRegs.XMM0 && loc <= X86MachRegs.XMM7
+				|| Float32Val.?(val)) {
+			movd_l_fval(frame, loc, val);
```

Comment:
What about moving the Float32Val case down into the match below? You'll need a Float64Val case too. It can be the case that the constant is null (indicating the default value for the type, which should be handled in the match too.